src/Symfony/Component/Mailer/Bridge/Sendgrid/Webhook/SendgridRequestParser.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Bridge\Sendgrid\Webhook;
+
+use EllipticCurve\Ecdsa;
+use EllipticCurve\PublicKey;
+use EllipticCurve\Signature;
+use Symfony\Component\HttpFoundation\ChainRequestMatcher;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestMatcher\IsJsonRequestMatcher;
+use Symfony\Component\HttpFoundation\RequestMatcher\MethodRequestMatcher;
+use Symfony\Component\HttpFoundation\RequestMatcherInterface;
+use Symfony\Component\Mailer\Bridge\Sendgrid\RemoteEvent\SendgridPayloadConverter;
+use Symfony\Component\RemoteEvent\Event\Mailer\AbstractMailerEvent;
+use Symfony\Component\RemoteEvent\Exception\ParseException;
+use Symfony\Component\Webhook\Client\AbstractRequestParser;
+use Symfony\Component\Webhook\Exception\RejectWebhookException;
+
+/**
+ * @author WoutervanderLoop.nl <info@woutervanderloop.nl>
+ */
+final class SendgridRequestParser extends AbstractRequestParser
+{
+    public function __construct(
+        private readonly SendgridPayloadConverter $converter,
+        private readonly bool $validateSignature,
+    ) {
+    }
+
+    protected function getRequestMatcher(): RequestMatcherInterface
+    {
+        return new ChainRequestMatcher([
+            new MethodRequestMatcher('POST'),
+            new IsJsonRequestMatcher(),
+        ]);
+    }
+
+    protected function doParse(Request $request, string $secret): ?AbstractMailerEvent
+    {
+        $content = $request->toArray();
+        if (
+            !isset($content[0]['email'])
+            || !isset($content[0]['timestamp'])
+            || !isset($content[0]['event'])
+            || !isset($content[0]['sg_message_id'])
+        ) {
+            throw new RejectWebhookException(406, 'Payload is malformed.');
+        }
+
+        if ($this->validateSignature) {
+            if (!$request->headers->get('X-Twilio-Email-Event-Webhook-Signature')
+                || !$request->headers->get('X-Twilio-Email-Event-Webhook-Timestamp')
+            ) {
+                throw new RejectWebhookException(406, 'Signature is required.');
+            }
+
+            if (!class_exists(Ecdsa::class)) {
+                throw new \LogicException('Package "starkbank/ecdsa" is required to use the "event-webhook-security" feature. Try running "composer require starkbank/ecdsa".');
```

Comment:
Why using this external library instead of relying on `openssl_verify` to do the job ?

Reply:
based on how sendgrid validates there requests in there own lib: https://github.com/sendgrid/sendgrid-php/blob/main/lib/eventwebhook/EventWebhook.php

Reply:
The sendgrid SDK is free to choose their dependencies as they want. But in Symfony, we are careful about them (and especially when it comes to a library that re-implements a crypto algorithm on its own with very few tests)

Reply:
@stof  I really cant find a way to make it work with default PHP openssl_verify. 

I'm not that experianced with crypto algorithms, but my guess is, that if a big platform a sendgrid choose a algorithm and library as this, it should be secure? :shrug: 

they use it in all there SDKs, (https://docs.sendgrid.com/for-developers/tracking-events/getting-started-event-webhook-security-features) 

Even [ecdsa-php](https://github.com/starkbank/ecdsa-php) has more than [11 million](https://packagist.org/packages/starkbank/ecdsa) installs. 



> (and especially when it comes to a library that re-implements a crypto algorithm on its own with very few tests).

but i completly understand your reason from your point of view. but I also think at the end of the story, the enduser/developer is responsible for making the descission wheter or not implement this feature, at the end, the install of the ecdsa lib is still optional

Reply:
I'm not saying that the ECDSA algorithm is bad (it is not). I'm saying that the ecdsa-php library implementing that algorithm in pure PHP is not properly tested (its tests are only covering 1 input for the crypto algorithm).
I trust a lot more the ECDSA implementation of openssl than of that library.

Having 11 million installs is probably due only to the sendgrid SDK which depends on it (this SDK has 24 million installs).
So this only says that the sendgrid team decided to use that implementation.

Reply:
The openssl implementation would look like that (my code snippet misses the error handling for openssl failures):

```php
// Sendgrid provides the verification key as base64-encoded DER data. Openssl wants a PEM format, which is a multiline version of the base64 data.
$pemKey = "-----BEGIN PUBLIC KEY-----\n".chunk_split($secret, 64, "\n")."-----END PUBLIC KEY-----\n";

$publicKey = openssl_get_publickey($pemKey);

return openssl_verify($timestampedPayload, base64_decode($signature), $publicKey, OPENSSL_ALGO_SHA256);
```

Reply:
Done